### ISSUE TITLE
Docs: Clearly mark rules that are frozen

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#18694]
### Docs page(s)

https://eslint.org/docs/latest/rules/

And each individual rule page

### What documentation issue do you want to solve?

While we say that stylistic rules are frozen, we don't clearly indicate which rules are frozen.

### What do you think is the correct solution?

We should clearly mark rules that are frozen with ❄️ in the place of the green checkbox on the rules index page.

We should also add a notice to the top of each rule page that is frozen, in the same style as "recommended", that says. "This rule is currently frozen and is not accepting changes."

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by kirkwaiblinger** [comment]
At typescript-eslint we've had success in turning down feature requests with no hard feelings due to similar notices.
Just mentioning in case it's helpful as an example, see https://github.com/typescript-eslint/typescript-eslint/issues/9554, https://typescript-eslint.io/rules/naming-convention/.

**Comment by amareshsm** [comment]
I would like to work on this. Assigning it to myself.

**Comment by nzakas** [comment]
@amareshsm thanks!

**Comment by mdjermanovic** [comment]
Where shall we maintain info about which rules are frozen? Maybe in `meta.docs.frozen` property (boolean) of rules?

**Comment by amareshsm** [comment]
> Where shall we maintain info about which rules are frozen? Maybe in `meta.docs.frozen` property (boolean) of rules?

Yes, I was thinking the same. In the meta.docs.frozen section, we can specify whether the rule is frozen or not.

**Comment by amareshsm** [comment]
I’m putting together a list of stylistic elements. If there’s anything missing, please let me know. If we already have an updated list, share the reference.

```
const frozenRules = [
    'array-bracket-newline',
    'array-bracket-spacing',
    'array-element-newline',
    'arrow-parens',
    'arrow-spacing',
    'block-spacing',
    'brace-style',
    'comma-dangle',
    'comma-spacing',
    'comma-style',
    'computed-property-spacing',
    'dot-location',
    'eol-last',
    'func-call-spacing',
    'function-call-argument-newline',
    'function-paren-newline',
    'generator-star-spacing',
    'implicit-arrow-linebreak',
    'indent',
    'jsx-quotes',
    'key-spacing',
    'keyword-spacing',
    'linebreak-style',
    'lines-between-class-members',
    'lines-around-comment',
    'lines-around-directive',
    'max-len',
    'max-statements-per-line',
    'multiline-ternary',
    'new-parens',
    'newline-after-var',
    'newline-before-return',
    'newline-per-chained-call',
    'no-confusing-arrow',
    'no-extra-parens',
    'no-extra-semi',
    'no-floating-decimal',
    'no-mixed-operators',
    'no-mixed-spaces-and-tabs',
    'no-multi-spaces',
    'no-multiple-empty-lines',
    'no-tabs',
    'no-trailing-spaces',
    'no-unexpected-multiline',
    'no-whitespace-before-property',
    'nonblock-statement-body-position',
    'object-curly-newline',
    'object-curly-spacing',
    'object-property-newline',
    'one-var-declaration-per-line',
    'operator-linebreak',
    'padded-blocks',
    'padding-line-between-statements',
    'prefer-destructuring',
    'quote-props',
    'quotes',
    'rest-spread-spacing',
    'semi',
    'semi-spacing',
    'semi-style',
    'space-before-blocks',
    'space-before-function-paren',
    'space-in-parens',
    'space-infix-ops',
    'space-unary-ops',
    'spaced-comment',
    'switch-colon-spacing',
    'template-curly-spacing',
    'template-tag-spacing',
    'wrap-iife',
    'wrap-regex',
    'yield-star-spacing'
];
```

**Comment by mdjermanovic** [comment]
Those are deprecated rules. What we need is to define a list of non-deprecated rules that are frozen.

**Comment by mdjermanovic** [comment]
To start, here's the list of all non-deprecated rules with `"type": "suggestion"` or `"type": "layout"`. I'm assuming that neither of `"type": "problem"` rules are frozen.


| **Rule** | **Frozen** |
| :---: | :---: |
| accessor-pairs | |
| arrow-body-style | |
| block-scoped-var | |
| camelcase | |
| capitalized-comments | |
| class-methods-use-this | |
| complexity | |
| consistent-return | |
| consistent-this | |
| curly | |
| default-case | |
| default-case-last | |
| default-param-last | |
| dot-notation | |
| eqeqeq | |
| func-name-matching | |
| func-names | |
| func-style | |
| grouped-accessor-pairs | |
| guard-for-in | |
| id-denylist | |
| id-length | |
| id-match | |
| init-declarations | |
| logical-assignment-operators | |
| max-classes-per-file | |
| max-depth | |
| max-lines | |
| max-lines-per-function | |
| max-nested-callbacks | |
| max-params | |
| max-statements | |
| new-cap | |
| no-alert | |
| no-array-constructor | |
| no-bitwise | |
| no-caller | |
| no-case-declarations | |
| no-console | |
| no-continue | |
| no-delete-var | |
| no-div-regex | |
| no-else-return | |
| no-empty | |
| no-empty-function | |
| no-empty-static-block | |
| no-eq-null | |
| no-eval | |
| no-extend-native | |
| no-extra-bind | |
| no-extra-boolean-cast | |
| no-extra-label | |
| no-global-assign | |
| no-implicit-coercion | |
| no-implicit-globals | |
| no-implied-eval | |
| no-inline-comments | |
| no-invalid-this | |
| no-iterator | |
| no-label-var | |
| no-labels | |
| no-lone-blocks | |
| no-lonely-if | |
| no-loop-func | |
| no-magic-numbers | |
| no-multi-assign | |
| no-multi-str | |
| no-negated-condition | |
| no-nested-ternary | |
| no-new | |
| no-new-func | |
| no-new-wrappers | |
| no-nonoctal-decimal-escape | |
| no-object-constructor | |
| no-octal | |
| no-octal-escape | |
| no-param-reassign | |
| no-plusplus | |
| no-proto | |
| no-redeclare | |
| no-regex-spaces | |
| no-restricted-exports | |
| no-restricted-globals | |
| no-restricted-imports | |
| no-restricted-properties | |
| no-restricted-syntax | |
| no-return-assign | |
| no-script-url | |
| no-sequences | |
| no-shadow | |
| no-shadow-restricted-names | |
| no-ternary | |
| no-throw-literal | |
| no-undef-init | |
| no-undefined | |
| no-underscore-dangle | |
| no-unneeded-ternary | |
| no-unused-expressions | |
| no-unused-labels | |
| no-useless-call | |
| no-useless-catch | |
| no-useless-computed-key | |
| no-useless-concat | |
| no-useless-constructor | |
| no-useless-escape | |
| no-useless-rename | |
| no-useless-return | |
| no-var | |
| no-void | |
| no-warning-comments | |
| no-with | |
| object-shorthand | |
| one-var | |
| operator-assignment | |
| prefer-arrow-callback | |
| prefer-const | |
| prefer-destructuring | |
| prefer-exponentiation-operator | |
| prefer-named-capture-group | |
| prefer-numeric-literals | |
| prefer-object-has-own | |
| prefer-object-spread | |
| prefer-promise-reject-errors | |
| prefer-regex-literals | |
| prefer-rest-params | |
| prefer-spread | |
| prefer-template | |
| radix | |
| require-await | |
| require-unicode-regexp | |
| require-yield | |
| sort-imports | |
| sort-keys | |
| sort-vars | |
| strict | |
| symbol-description | |
| unicode-bom | |
| vars-on-top | |
| yoda | |

**Comment by mdjermanovic** [comment]
I marked the rules I think should be frozen in the list above. @eslint/eslint-tsc thoughts?

**Comment by fasttime** [comment]
Here are a few more rules that were deemed stylistic at some point. I'm not 100% sure about all of them, but we should at least discuss them:
- [func-names](https://github.com/eslint/eslint/issues/16182)
- [max-lines](https://github.com/eslint/eslint/issues/16195)
- [max-lines-per-function](https://github.com/eslint/eslint/issues/18786)
- [max-params](https://github.com/eslint/eslint/pull/17796)
- [max-statements](https://github.com/eslint/eslint/issues/18651)
- [new-cap](https://github.com/eslint/eslint/issues/15724)
- [no-bitwise](https://github.com/eslint/eslint/pull/18205)
- [no-new](https://github.com/eslint/eslint/issues/15705)
- [no-param-reassign](https://github.com/eslint/eslint/issues/15628)
- [no-shadow](https://github.com/eslint/eslint/issues/17233)
- [no-useless-computed-key](https://github.com/eslint/eslint/issues/17536)
- [no-warning-comments](https://github.com/eslint/eslint/issues/15133)

**Comment by fasttime** [comment]
According to the rule description, [`block-scoped-var`](https://eslint.org/docs/latest/rules/block-scoped-var) will _help newcomers avoid difficult bugs with variable hoisting._ I think that doesn't qualify as stylistic.

I'm also not sure about [`prefer-object-has-own`](https://eslint.org/docs/latest/rules/prefer-object-has-own). This rule targets the case when `Object.prototype` is unavailable (non-stylistic), and also recommends a shorter alternative (stylistic).

**Comment by mdjermanovic** [comment]
I believe `prefer-object-has-own` only recommends a shorter equivalent, so I think it's only stylistic.

```js
/*eslint prefer-object-has-own: "error"*/

Object.prototype.hasOwnProperty.call(obj, "a"); // error

Object.hasOwnProperty.call(obj, "a"); // error

({}).hasOwnProperty.call(obj, "a"); // error

obj.hasOwnProperty("a"); // no error, this is reported by no-prototype-builtins

```

As for `block-scoped-var`, my reasoning was more along the lines of "we don't want to add additional options to this rule" as @nzakas said, because `var` is rarely used nowadays.

**Comment by mdjermanovic** [comment]
> Here are a few more rules that were deemed stylistic at some point. I'm not 100% sure about all of them, but we should at least discuss them:

Good finds that solidify the need to clearly define the list of frozen rules, for ourselves as well.

I'll just briefly explain my reasoning for omitting these in the first pass.

> * [func-names](https://github.com/eslint/eslint/issues/16182)

Per the docs, this helps in debugging / stack traces, so it seemed to me more than just stylistic.

> * [max-lines](https://github.com/eslint/eslint/issues/16195)
> * [max-lines-per-function](https://github.com/eslint/eslint/issues/18786)
> * [max-params](https://github.com/eslint/eslint/pull/17796)
> * [max-statements](https://github.com/eslint/eslint/issues/18651)

These rules measure the complexity of code (though by very simple metrics), so if the `complexity` rule is not frozen, it felt like these shouldn't be frozen too.

> * [new-cap](https://github.com/eslint/eslint/issues/15724)

I thought this rule might help find bugs - when a constructor is called without `new` or a non-constructor is called with `new`-  but from the linked issue I recall we consider it just stylistic (enforcing naming conventions), so I'm fine with marking it as frozen.

> * [no-bitwise](https://github.com/eslint/eslint/pull/18205)

Per the docs, it seemed to me this rule helps find possible errors (e.g., mistyped `&&` or `||`).

> * [no-new](https://github.com/eslint/eslint/issues/15705)

I thought this rule might help find bugs similar to those caught by `no-unused-expression`, but per the docs and the error message it seems the intent of this rule is to disallow creating classes/constructors whose only purpose is to produce side effects, which seems as a stylistic preference so I'm fine with marking this rule as frozen.

> * [no-param-reassign](https://github.com/eslint/eslint/issues/15628)

While the docs describe it as stylistic, and the problem with `arguments` is now mostly obsolete, I recall there were cases where this rule prevented errors in the code (e.g., https://github.com/eslint/eslint/issues/15244). 

> * [no-shadow](https://github.com/eslint/eslint/issues/17233)

While it can be argued that this is a stylistic preference, it's a common practice (not just in JS) that also prevents future errors. I think this rule should stay non-frozen.

> * [no-useless-computed-key](https://github.com/eslint/eslint/issues/17536)

I think I just missed this one, fine with marking it as frozen.

> * [no-warning-comments](https://github.com/eslint/eslint/issues/15133)

This felt like a generic configurable rule, like e.g., no-restricted-imports. I'm fine with marking it as frozen though.

**Comment by fasttime** [comment]
> We don't necessarily need to say that a rule is stylistic for it to be frozen (though that's where this discussion started). I think it's perfectly fine to say "we don't want to add additional options to `block-scoped-var`" regardless of whether we can agree if it's stylistic or not.
> 
> Similar with `prefer-object-has-own` -- I don't think this is stylistic but I honestly don't know why we'd want to make any changes to it.

Thanks for the clarifications. I've seen sometimes users requesting to add suggestions to a rule, or to make the problems reported by a rule more granular, or to improve the message or location where a problem is reported. My impression is that these changes should not be accepted for frozen rules, but they could be still useful for other rules that are not frozen. So by declaring a rule frozen we will miss not only requests to add an option (which may not be appropriate for certain rules like `prefer-object-has-own`), but also other improvements that could be helpful.

I would like to understand better what rules we would like to be frozen apart from the stylistic ones.

**Comment by nzakas** [comment]
Folks, I think we're overthinking this. Whether a rule is frozen or not doesn't have to be set in stone. Let's just get started with a list that makes sense and go from there. We can always unfreeze or refreeze rules later.

Where there's consensus on frozen rules, let's start with those. We can then bring up individual rules for consideration separate from this effort.

**Comment by fasttime** [comment]
> Then, we could start with the list from [#18694 (comment)](https://github.com/eslint/eslint/issues/18694#issuecomment-2355493603), minus `block-scoped-var` and `prefer-object-has-own`?

Sounds good :+1: I'd be fine with marking also `no-useless-computed-key` and `no-warning-comments` as frozen is there are no concerns from the team.

**Comment by nzakas** [comment]
> Sounds good 👍 I'd be fine with marking also `no-useless-computed-key` and `no-warning-comments` as frozen is there are no concerns from the team.

Works for me.

**Comment by mdjermanovic** [comment]
Works for me too. I updated the list in https://github.com/eslint/eslint/issues/18694#issuecomment-2355493603:

* Removed `block-scoped-var` and `prefer-object-has-own`.
* Added `no-useless-computed-key` and `no-warning-comments`

**Comment by fasttime** [comment]
Thoughts about adding `no-useless-concat` to the list of frozen rules? Issue #19113 raises the question of whether that rule is frozen or not, and I couldn't find any relevant previous discussions. I think it should be by our standards because it addresses a stylistic issue.

**Comment by nzakas** [comment]
I think it's fine to add `no-useless-concat`. After that, though, I think we should "freeze" this list and get to work implementing the changes on the website.

**Comment by mdjermanovic** [comment]
I also agree with adding `no-useless-concat` to the list of frozen rules. I've updated the list now.

**Comment by fasttime** [comment]
@amareshsm The list of frozen rules in https://github.com/eslint/eslint/issues/18694#issuecomment-2355493603 is ready, a PR is welcome.

**Comment by Tanujkanti4441** [comment]
Can i take this if no one is working?

**Comment by amareshsm** [comment]
> Can I take this if no one else is working on it for now?

I have done some I will send a PR today. If any major is required from there you can it forward.
